wrapper/src/main/java/software/amazon/jdbc/util/monitoring/MonitorService.java

```diff
@@ -24,52 +24,64 @@
+   * @param monitorClass         the class of the monitor, eg `CustomEndpointMonitorImpl.class`.
+   * @param timeToLiveNanos      how long a monitor should be stored before being considered expired, in nanoseconds. If
+   *                             the monitor is expired and shouldDisposeFunc returns `true`, the monitor will be
+   *                             stopped.
+   * @param inactiveTimeoutNanos a duration in nanoseconds defining the maximum amount of time that a monitor should
+   *                             take between updating its last-updated timestamp. If a monitor has not updated its
+   *                             last-updated timestamp within this value it will be considered stuck.
+   * @param errorResponses       a Set defining actions to take if the monitor is in an error state.
+   * @param shouldDisposeFunc    a function defining whether an item should be stopped if expired. If `null` is
+   *                             passed, the monitor will always be stopped if the monitor is expired.
    */
-  void registerMonitorTypeIfAbsent(
-      String monitorType,
+  <T extends Monitor> void registerMonitorTypeIfAbsent(
+      Class<T> monitorClass,
+      long timeToLiveNanos,
+      long inactiveTimeoutNanos,
       Set<MonitorErrorResponse> errorResponses,
-      long expirationTimeNs,
-      @Nullable ShouldDisposeFunc<Monitor> shouldDisposeFunc);
+      @Nullable ShouldDisposeFunc<T> shouldDisposeFunc);
 
   /**
    * Creates and starts the given monitor if it does not already exist and stores it under the given monitor type and
-   * key.
+   * key. If the monitor already exists, its time-to-live duration will be renewed, even if it was already expired.
    *
-   * @param monitorType     a String representing the monitor type, eg "customEndpoint".
+   * @param monitorClass    the class of the monitor, eg `CustomEndpointMonitorImpl.class`.
    * @param key             the key for the monitor, eg
    *                        "custom-endpoint.cluster-custom-XYZ.us-east-2.rds.amazonaws.com:5432".
-   * @param monitorSupplier an initialization lambda that can be used to create the monitor if it is absent.
+   * @param monitorSupplier a supplier lambda that can be used to create the monitor if it is absent.
    */
-  void runIfAbsent(
-      String monitorType,
-      Object key,
-      Supplier<Monitor> monitorSupplier);
+  <T extends Monitor> void runIfAbsent(Class<T> monitorClass, Object key, Supplier<T> monitorSupplier);
+
+  /**
+   * Process a monitor error. The monitor service will respond to the error based on the monitor error responses defined
+   * when the monitor type was registered.
+   *
+   * @param monitor   the monitor that encountered the unexpected exception.
+   * @param key       the key for the monitor.
+   * @param exception the unexpected exception that occurred.
+   */
+  void processMonitorError(Monitor monitor, Object key, Exception exception);
```

Comment:
I added this method on the suggestion that monitors be able to notify the monitor service when they encounter an unexpected exception. The monitor service can then respond to this based on pre-configured settings for that monitor type, for example it may try and re-create the monitor. I think this is a good idea, but it requires the monitor to pass in its own key here, so that the monitor service can update its cache with the re-created monitor. It seems bad that the monitor needs to keep track of its own key, but I'm not sure how to fix this.

Reply:
You're passing monitor reference to monitor service. I believe that's sufficient. you don't need a key parameter. Monitor service can find a monitor and all related info by monitor reference.

Reply:
Yeah, the monitor service will have to iterate over its entries to find it but I guess that doesn't matter much since there shouldn't be a large number of monitors of any given type